File: UniPy/Assets/Scripts/Utils/Logger/Log.cs
namespace Disc0ver.Engine
{
    public static class Log
    {
        private static ILogImpl _logImpl = new DefaultLogImpl();

        public static void SetLogImpl(ILogImpl logImpl)
        {
            _logImpl = logImpl;
        }

        public static void Debug(object message)
        {
            _logImpl?.Log(LogLevel.Debug, message);
        }

        public static void Info(object message)
        {
            _logImpl?.Log(LogLevel.Info, message);
        }

        public static void Error(object message)
        {
            _logImpl?.Log(LogLevel.Error, message);
        }

        public static void Warning(object message)
        {
            _logImpl?.Log(LogLevel.Warning, message);
        }

        public static void Fatal(object message)
        {
            _logImpl?.Log(LogLevel.Fatal, message);
        }
    }
}

File: UniPy/Python/PyScripts/Utils/Log/Log.py
class Log(object):

    def __init__(self) -> None:
        self._logImpl = None

    def SetLogImpl(self, logImpl):
        self._logImpl = logImpl

    def FormatMsg(self, msg):
        return f"[Python] - {msg}"

    def Debug(self, msg):
        new_msg = self.FormatMsg(msg)
        if self._logImpl:
            self._logImpl.Debug(new_msg)
        else:
            print(new_msg)
    
    def Info(self, msg):
        new_msg = self.FormatMsg(msg)
        if self._logImpl:
            self._logImpl.Info(new_msg)
        else:
            print(new_msg)
    
    def Error(self, msg):
        new_msg = self.FormatMsg(msg)
        if self._logImpl:
            self._logImpl.Error(new_msg)
        else:
            print(new_msg)

    def Warning(self, msg):
        new_msg = self.FormatMsg(msg)
        if self._logImpl:
            self._logImpl.Warning(new_msg)
        else:
            print(new_msg)
    
    def Fatal(self, msg):
        new_msg = self.FormatMsg(msg)
        if self._logImpl:
            self._logImpl.Fatal(new_msg)
        else:
            print(new_msg)
